In [1]:
import pandas as pd

In [2]:
df_listings = pd.read_csv("../data/Listing/listed_transactions.csv")
df_listings.head()

C:\Users\mayab\AppData\Local\Temp\ipykernel_27328\3410094314.py:1: DtypeWarning: Columns (0: ListAgentEmail, 1: BuyerAgencyCompensationType) have mixed types. Specify dtype option on import or set low_memory=False.
  df_listings = pd.read_csv("../data/Listing/listed_transactions.csv")


,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,...,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1,Year
0,90000.0,1075010398,miriamlara03@gmail.com,NaN,NaN,Miriam,Lara,34.097939,-117.909653,1045 N Azusa 61,...,NaN,2.0,Covina Valley Unified,91722,NaN,0.0,NaN,NaN,1045 N Azusa 61,2024
1,1500000.0,1074974457,janelle@judsonre.com,NaN,NaN,Janelle,Judson,33.121241,-117.081614,NaN,...,NaN,NaN,NaN,92025,NaN,0.0,NaN,NaN,NaN,2024
2,1340000.0,1074973329,haleh360@Gmail.com,NaN,NaN,Haleh,Dowlatshahi,34.052207,-118.408445,2220 Avenue Of The Stars 2704,...,False,NaN,NaN,90067,NaN,2105.0,177861.0,NaN,2220 Avenue Of The Stars 2704,2024
3,2500000.0,1074954552,Reneechen@yourhomesoldguaranteed.com,NaN,NaN,Renee,Chen,33.496363,-117.691677,16 Palisades,...,False,3.0,Capistrano Unified,92677,NaN,254.0,5300.0,NaN,16 Palisades,2024
4,3150000.0,1074936537,anader@dppre.com,NaN,NaN,Margaret,Nader,34.119345,-118.111254,1615 Waverly Road,...,NaN,2.0,NaN,91108,NaN,NaN,9404.0,NaN,1615 Waverly Road,2024


In [3]:
df_sold = pd.read_csv("../data/Sold/sold_transactions.csv")
df_sold.head()

C:\Users\mayab\AppData\Local\Temp\ipykernel_27328\4086709524.py:1: DtypeWarning: Columns (0: BuyerAgentAOR, 1: ListAgentAOR, 2: WaterfrontYN, 3: ListAgentEmail, 4: OriginatingSystemName, 5: OriginatingSystemSubName, 6: BuyerAgencyCompensationType, 7: latfilled, 8: lonfilled) have mixed types. Specify dtype option on import or set low_memory=False.
  df_sold = pd.read_csv("../data/Sold/sold_transactions.csv")


,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,...,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,BuyerAgencyCompensation,latfilled,lonfilled,Year
0,Mlslistings,Mlslistings,"Carpet,Tile,Wood",True,NaN,NaN,False,499000.0,551985747,jwachter@cbnorcal.com,...,6472.0,NaN,NaN,CRMLS,CRMLS_MLSL,NaN,NaN,NaN,NaN,2024
1,HighDesert,HighDesert,NaN,NaN,NaN,NaN,NaN,0.0,535486633,eabrown@lee-associates.com,...,NaN,52320.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN,2024
2,OrangeCounty,OrangeCounty,NaN,True,NaN,NaN,NaN,75000.0,529986282,Joe@9WINWIN.com,...,NaN,217364.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN,2024
3,InlandValleys,InlandValleys,NaN,True,NaN,NaN,NaN,199000.0,529618166,carolthefinder@yahoo.com,...,NaN,217800.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN,2024
4,SouthwestRiversideCounty,SouthwestRiversideCounty,NaN,True,NaN,NaN,NaN,19500.0,522614340,jtavisola@tavisola.com,...,0.0,108883.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN,2024


##### Number of Rows and Columns

In [4]:
df_listings.shape

(853754, 85)

In [5]:
df_sold.shape

(591115, 85)

For the listing transactions, there are 853754 rows and 85 columns (original is 84). With the sold transactions, there 591115 rows and 85 columns (also originally 84).

##### Review column data types

In [6]:
df_listings.info()

<class 'pandas.DataFrame'>
RangeIndex: 853754 entries, 0 to 853753
Data columns (total 85 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   OriginalListPrice             850444 non-null  float64
 1   ListingKey                    853754 non-null  int64  
 2   ListAgentEmail                780379 non-null  str    
 3   CloseDate                     255893 non-null  str    
 4   ClosePrice                    233059 non-null  float64
 5   ListAgentFirstName            848740 non-null  str    
 6   ListAgentLastName             853678 non-null  str    
 7   Latitude                      741807 non-null  float64
 8   Longitude                     742511 non-null  float64
 9   UnparsedAddress               851597 non-null  str    
 10  PropertyType                  853754 non-null  str    
 11  LivingArea                    747280 non-null  float64
 12  ListPrice                     851619 non-null  float64


In [7]:
df_sold.info()

<class 'pandas.DataFrame'>
RangeIndex: 591115 entries, 0 to 591114
Data columns (total 85 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   BuyerAgentAOR                 519608 non-null  str    
 1   ListAgentAOR                  523069 non-null  str    
 2   Flooring                      347464 non-null  str    
 3   ViewYN                        531925 non-null  object 
 4   WaterfrontYN                  322 non-null     object 
 5   BasementYN                    9738 non-null    object 
 6   PoolPrivateYN                 516977 non-null  object 
 7   OriginalListPrice             589404 non-null  float64
 8   ListingKey                    591115 non-null  int64  
 9   ListAgentEmail                547714 non-null  str    
 10  CloseDate                     591115 non-null  str    
 11  ClosePrice                    591108 non-null  float64
 12  ListAgentFirstName            587682 non-null  str    


#### **Missing Value Analysis**

##### Calculate missing counts and percentages per column

##### *Listing*

In [8]:
missing_listing_counts = df_listings.isnull().sum()
missing_listing_counts

OriginalListPrice                 3310
ListingKey                           0
ListAgentEmail                   73375
CloseDate                       597861
ClosePrice                      620695
                                 ...  
AssociationFee                  275935
LotSizeSquareFeet                78866
MiddleOrJuniorSchoolDistrict    853754
UnparsedAddress.1                 2157
Year                                 0
Length: 85, dtype: int64

In [9]:
missing_listing_percent = (df_listings.isnull().mean()) * 100
missing_listing_percent

OriginalListPrice                 0.387700
ListingKey                        0.000000
ListAgentEmail                    8.594396
CloseDate                        70.027315
ClosePrice                       72.701856
                                   ...    
AssociationFee                   32.320200
LotSizeSquareFeet                 9.237556
MiddleOrJuniorSchoolDistrict    100.000000
UnparsedAddress.1                 0.252649
Year                              0.000000
Length: 85, dtype: float64

In [10]:
missing_summary = pd.DataFrame({
    "missing_listing_counts": missing_listing_counts,
    "missing_listing_percent": missing_listing_percent
})

In [11]:
missing_listing_summary = missing_summary.sort_values(by="missing_listing_percent", ascending=False)
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
MiddleOrJuniorSchoolDistrict                  853754                    100.0
ElementarySchoolDistrict                      853754                    100.0
CoveredSpaces                                 853754                    100.0
AboveGradeFinishedArea                        853754                    100.0
FireplacesTotal                               853754                    100.0
...                                              ...                      ...
PropertyType.1                                     0                      0.0
ListingKeyNumeric                                  0                      0.0
DaysOnMarket.1                                     0                      0.0
ListingContractDate                                0                      0.0
Year                                               0                      0.0

[85 rows x 2 columns]


In [12]:
missing_listing_summary = missing_listing_summary[missing_listing_summary["missing_listing_counts"] > 0]
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
MiddleOrJuniorSchoolDistrict                  853754               100.000000
ElementarySchoolDistrict                      853754               100.000000
CoveredSpaces                                 853754               100.000000
AboveGradeFinishedArea                        853754               100.000000
FireplacesTotal                               853754               100.000000
...                                              ...                      ...
PostalCode                                       208                 0.024363
ListAgentLastName                                 76                 0.008902
ListAgentLastName.1                               76                 0.008902
StateOrProvince                                   66                 0.007731
CountyOrParish                                     1                 0.000117

[74 rows x 2 columns]


In [13]:
missing_listing_summary["missing_listing_percent"] = missing_listing_summary["missing_listing_percent"].round(2)
print(missing_listing_summary)

                              missing_listing_counts  missing_listing_percent
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
FireplacesTotal                               853754                   100.00
...                                              ...                      ...
PostalCode                                       208                     0.02
ListAgentLastName                                 76                     0.01
ListAgentLastName.1                               76                     0.01
StateOrProvince                                   66                     0.01
CountyOrParish                                     1                     0.00

[74 rows x 2 columns]


In [14]:
missing_above_90 = missing_listing_summary[missing_listing_summary['missing_listing_percent'] > 90]
print(missing_above_90)

                              missing_listing_counts  missing_listing_percent
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
FireplacesTotal                               853754                   100.00
TaxYear                                       852886                    99.90
BelowGradeFinishedArea                        850334                    99.60
BusinessType                                  847779                    99.30
TaxAnnualAmount                               847661                    99.29
CoBuyerAgentFirstName                         834049                    97.69
BuilderName                                   824452                    96.57
LotSizeDimensions                             805053            

Dropping Variables:
- MiddleOrJuniorSchoolDistrict                  
- ElementarySchoolDistrict
- CoveredSpaces
- AboveGradeFinishedArea
- FireplacesTotal
- TaxYear
- BelowGradeFinishedArea
- BusinessType
- TaxAnnualAmount
- CoBuyerAgentFirstName
- BuilderName
- LotSizeDimensions
- ElementarySchool
- MiddleOrJuniorSchool

In [15]:
missing_above_70 = missing_listing_summary[missing_listing_summary['missing_listing_percent'] > 70]
print(missing_above_70)

                              missing_listing_counts  missing_listing_percent
MiddleOrJuniorSchoolDistrict                  853754                   100.00
ElementarySchoolDistrict                      853754                   100.00
CoveredSpaces                                 853754                   100.00
AboveGradeFinishedArea                        853754                   100.00
FireplacesTotal                               853754                   100.00
TaxYear                                       852886                    99.90
BelowGradeFinishedArea                        850334                    99.60
BusinessType                                  847779                    99.30
TaxAnnualAmount                               847661                    99.29
CoBuyerAgentFirstName                         834049                    97.69
BuilderName                                   824452                    96.57
LotSizeDimensions                             805053            

In [16]:
missing_above_70.shape

(31, 2)

In [17]:
missing_above_70

,missing_listing_counts,missing_listing_percent
MiddleOrJuniorSchoolDistrict,853754,100.00
ElementarySchoolDistrict,853754,100.00
CoveredSpaces,853754,100.00
AboveGradeFinishedArea,853754,100.00
FireplacesTotal,853754,100.00
TaxYear,852886,99.90
BelowGradeFinishedArea,850334,99.60
BusinessType,847779,99.30
TaxAnnualAmount,847661,99.29
CoBuyerAgentFirstName,834049,97.69


To perserve the variables, it would be sufficient if each variable contain less than 70% of missing values in order to use imputation to replace the missingness in the data. 

So, variables with 70% of missing variables will be removed from the dataset because through imputation with variables with a significant number of missingness, it may contribute to bias towards certain values than others. 

Keep these key variables (from above 70):
- ClosePrice

In [19]:
df_listing_clean = df_listings.drop(columns=['MiddleOrJuniorSchoolDistrict', 'ElementarySchoolDistrict', 'CoveredSpaces',
                                             'AboveGradeFinishedArea', 'FireplacesTotal', 'TaxYear', 
                                             'BelowGradeFinishedArea', 'BusinessType', 'TaxAnnualAmount',
                                             'CoBuyerAgentFirstName', 'BuilderName', 'LotSizeDimensions',
                                             'ElementarySchool', 'MiddleOrJuniorSchool', 'HighSchool',
                                             'BuyerAgencyCompensation', 'BuyerAgencyCompensationType', 'BuildingAreaTotal',
                                             'CoListAgentFirstName', 'CoListAgentLastName', 'CoListOfficeName', 'BuyerOfficeAOR',
                                             'BuyerOfficeName', 'BuyerOfficeName.1', 'AssociationFeeFrequency', 'BuyerAgentFirstName',
                                             'BuyerAgentMlsId', 'BuyerAgentLastName', 'CloseDate', 'CloseDate.1'])

In [20]:
df_listing_clean.shape

(853754, 55)

##### *Sold*

In [21]:
missing_sold_counts = df_sold.isnull().sum()
missing_sold_counts

BuyerAgentAOR                   71507
ListAgentAOR                    68046
Flooring                       243651
ViewYN                          59190
WaterfrontYN                   590793
                                ...  
BuyerAgencyCompensationType    523368
BuyerAgencyCompensation        523335
latfilled                      495322
lonfilled                      495322
Year                                0
Length: 85, dtype: int64

In [22]:
missing_sold_percent = (df_sold.isnull().mean()) * 100
missing_sold_percent

BuyerAgentAOR                  12.096969
ListAgentAOR                   11.511466
Flooring                       41.218883
ViewYN                         10.013280
WaterfrontYN                   99.945527
                                 ...    
BuyerAgencyCompensationType    88.539117
BuyerAgencyCompensation        88.533534
latfilled                      83.794524
lonfilled                      83.794524
Year                            0.000000
Length: 85, dtype: float64

In [23]:
missing_summary = pd.DataFrame({
    "missing_sold_counts": missing_sold_counts,
    "missing_sold_percent": missing_sold_percent
})

In [24]:
missing_sold_summary = missing_summary.sort_values(by="missing_sold_percent", ascending=False)
print(missing_sold_summary)

                              missing_sold_counts  missing_sold_percent
MiddleOrJuniorSchoolDistrict               591115                 100.0
ElementarySchoolDistrict                   591115                 100.0
CoveredSpaces                              591115                 100.0
FireplacesTotal                            591115                 100.0
AboveGradeFinishedArea                     591115                 100.0
...                                           ...                   ...
CountyOrParish                                  0                   0.0
ListOfficeName                                  0                   0.0
MlsStatus                                       0                   0.0
StateOrProvince                                 0                   0.0
Year                                            0                   0.0

[85 rows x 2 columns]


In [25]:
missing_sold_summary = missing_sold_summary[missing_sold_summary["missing_sold_counts"] > 0]
print(missing_sold_summary)

                              missing_sold_counts  missing_sold_percent
MiddleOrJuniorSchoolDistrict               591115            100.000000
ElementarySchoolDistrict                   591115            100.000000
CoveredSpaces                              591115            100.000000
FireplacesTotal                            591115            100.000000
AboveGradeFinishedArea                     591115            100.000000
...                                           ...                   ...
PostalCode                                    165              0.027913
ListAgentFullName                             161              0.027237
ListingContractDate                            83              0.014041
ListAgentLastName                              61              0.010319
ClosePrice                                      7              0.001184

[74 rows x 2 columns]


In [26]:
missing_sold_summary["missing_sold_percent"] = missing_sold_summary["missing_sold_percent"].round(2)
print(missing_sold_summary)

                              missing_sold_counts  missing_sold_percent
MiddleOrJuniorSchoolDistrict               591115                100.00
ElementarySchoolDistrict                   591115                100.00
CoveredSpaces                              591115                100.00
FireplacesTotal                            591115                100.00
AboveGradeFinishedArea                     591115                100.00
...                                           ...                   ...
PostalCode                                    165                  0.03
ListAgentFullName                             161                  0.03
ListingContractDate                            83                  0.01
ListAgentLastName                              61                  0.01
ClosePrice                                      7                  0.00

[74 rows x 2 columns]


In [27]:
missing_above_90 = missing_sold_summary[missing_sold_summary['missing_sold_percent'] > 90]
print(missing_above_90)

                              missing_sold_counts  missing_sold_percent
MiddleOrJuniorSchoolDistrict               591115                100.00
ElementarySchoolDistrict                   591115                100.00
CoveredSpaces                              591115                100.00
FireplacesTotal                            591115                100.00
AboveGradeFinishedArea                     591115                100.00
WaterfrontYN                               590793                 99.95
TaxYear                                    590748                 99.94
BusinessType                               589483                 99.72
TaxAnnualAmount                            588608                 99.58
BelowGradeFinishedArea                     588525                 99.56
BasementYN                                 581377                 98.35
BuilderName                                568536                 96.18
LotSizeDimensions                          560102               

In [28]:
missing_above_70 = missing_sold_summary[missing_sold_summary['missing_sold_percent'] > 70]
print(missing_above_70)

                              missing_sold_counts  missing_sold_percent
MiddleOrJuniorSchoolDistrict               591115                100.00
ElementarySchoolDistrict                   591115                100.00
CoveredSpaces                              591115                100.00
FireplacesTotal                            591115                100.00
AboveGradeFinishedArea                     591115                100.00
WaterfrontYN                               590793                 99.95
TaxYear                                    590748                 99.94
BusinessType                               589483                 99.72
TaxAnnualAmount                            588608                 99.58
BelowGradeFinishedArea                     588525                 99.56
BasementYN                                 581377                 98.35
BuilderName                                568536                 96.18
LotSizeDimensions                          560102               

In [29]:
missing_above_70

,missing_sold_counts,missing_sold_percent
MiddleOrJuniorSchoolDistrict,591115,100.00
ElementarySchoolDistrict,591115,100.00
CoveredSpaces,591115,100.00
FireplacesTotal,591115,100.00
AboveGradeFinishedArea,591115,100.00
WaterfrontYN,590793,99.95
TaxYear,590748,99.94
BusinessType,589483,99.72
TaxAnnualAmount,588608,99.58
BelowGradeFinishedArea,588525,99.56


In [30]:
df_sold_clean = df_sold.drop(columns=['MiddleOrJuniorSchoolDistrict', 'ElementarySchoolDistrict', 'CoveredSpaces', 
                                      'FireplacesTotal', 'AboveGradeFinishedArea', 'WaterfrontYN',
                                      'TaxYear', 'BusinessType', 'TaxAnnualAmount', 'BelowGradeFinishedArea',
                                      'BasementYN', 'BuilderName', 'LotSizeDimensions', 'CoBuyerAgentFirstName',
                                      'OriginatingSystemName', 'OriginatingSystemSubName', 'ElementarySchool', 
                                      'MiddleOrJuniorSchool', 'BuyerAgencyCompensationType', 'BuyerAgencyCompensation', 
                                      'BuildingAreaTotal', 'HighSchool', 'latfilled', 'lonfilled', 'CoListAgentFirstName', 
                                      'CoListAgentLastName', 'CoListOfficeName', 'AssociationFeeFrequency'])

In [31]:
df_sold_clean.shape

(591115, 57)